src/controllers/EventsController.js

```diff
@@ -1,9 +1,177 @@
 // @flow
 
+import moment from 'moment';
 import Controller from './Controller';
+import CoreController from '../lib/CoreController';
+import route from '../decorators/route';
+import httpVerb from '../decorators/httpVerb';
+
+import v1Api from '../views/api_v1';
+import logger from '../lib/logger';
 
 class EventsController extends Controller {
+  _aliveInterval: ?string;
+  _lastMessage: ?Date;
+  _socket: ?Object;
+
+  _cleanup() {
+    try {
+      if (this._socket) {
+        this._socket.close();
+        this._socket = null;
+      }
+
+      if (this.response && this.response.socket) {
+        this.response.socket.end();
+        this.response.end();
+      }
+
+      if (this._aliveInterval) {
+        clearInterval(this._aliveInterval);
+        this._aliveInterval = null;
+      }
+    } catch (error) {
+      logger.error(`pipeEvents cleanup error: ${error}`);
+    }
+  }
+
+  _keepAlive() {
+    if (((new Date()) - this._lastMessage) >= 9000) {
+      this._lastMessage = new Date();
+      this.response.write('\n');
+      this._checkSocket();
+    }
+  }
+
+  _checkSocket() {
+    try {
+      if (!this._socket) {
+        this._cleanup();
+        return false;
+      }
+
+      if (this.response.socket.destroyed) {
+        logger.log('Socket destroyed, cleaning up Event listener');
+        this._cleanup();
+        return false;
+      }
+
+      return true;
+    } catch (error) {
+      logger.error(`pipeEvents - error checking socket ${error}`);
+      return false;
+    }
+  }
+
+  _writeEventGen() {
+    return (
+      name: string,
+      data: ?Object,
+      ttl: ?number,
+      publishedAt: ?Date,
+      coreid: ? string,
+    ) => {
+      // if (filterCoreId && (filterCoreId !== coreid)) {
+      //   return;
+      // }
+      if (!this._checkSocket()) {
+        return;
+      }
+
+      try {
+        this._lastMessage = new Date();
+
+        const eventData = {
+          coreid: coreid || null,
+          data: data || null,
+          published_at: publishedAt || null,
+          ttl: ttl || null,
+        };
+
+        this.response.write(`event: ${name} \n`);
+        this.response.write(`data: ${JSON.stringify(eventData)} \n\n`);
+      } catch (error) {
+        logger.error(`pipeEvents - write error: ${error}`);
+      }
+    };
+  }
+
+  _pipeEvents() {
+    try {
+      this.request.socket.setNoDelay();
+      console.log('headers sent', this.response.headersSent);
+      this.response.writeHead(200, {
```

Comment:
hm.. I get `UnhandledPromiseRejectionWarning: Unhandled promise
rejection (rejection id: 1): Error: Can't set headers after they
are sent.`, but haven't  figured out where is the problem. before this call headers are not sent, after they are sent...so the problem somewhere else..will dig into tommorow..also I'm not sure if we need these headers at all.